runners/flink/src/main/java/org/apache/beam/runners/flink/LyftFlinkStreamingPortableTranslations.java

```diff
@@ -357,6 +375,216 @@ private void translateKinesisInput(
+    // Add kinesis parallelism
+    JsonNode kinesisParallelism = userKinesisConfig.get("parallelism");
+    if (kinesisParallelism != null) {
+      builder = builder.withParallelism(kinesisParallelism.asInt());
+    }
+    // Add kinesis properties
+    Map<String, String> kinesisProps = mapper.convertValue(
+        userKinesisConfig.get("properties"), new TypeReference<Map<String, String>>() {});
+    if (kinesisProps != null) {
+      for (Map.Entry<String ,String> property : kinesisProps.entrySet()) {
+        properties.put(property.getKey(), property.getValue());
+      }
+      builder = builder.withProperties(properties);
+    }
+    // Add kinesis stream start mode
+    JsonNode streamStartMode = userKinesisConfig.get("stream_start_mode");
+    if (streamStartMode != null) {
+      builder = builder.withStreamStartMode(streamStartMode.asText());
+    }
+
+    return builder.build();
+  }
+
+  static class EventToWindowedValue implements MapFunction<Event, WindowedValue<byte[]>> {
+
+    @Override
+    public WindowedValue<byte[]> map(Event value) {
+      return WindowedValue.timestampedValueInGlobalWindow(
+          getBytes(value), new Instant(getTimestamp(value)));
+    }
+
+    private long getTimestamp(Event event) {
+      Object occurredAt = event.get(EventField.EventOccurredAt.fieldName());
+      long occurredAtMs = 0L;
+      try {
+        if (occurredAt instanceof String) {
+          occurredAtMs = parseDateTime((String) occurredAt);
+        } else if (occurredAt instanceof Long || occurredAt instanceof Integer) {
+          occurredAtMs = (long) occurredAt;
+        } else if (occurredAt instanceof Timestamp) {
+          occurredAtMs = ((Timestamp) occurredAt).getTime();
+        }
+        else {
+          LOG.warn("Occurred at ts unrecognized");
+        }
+        if (event.contains(EventField.EventLoggedAt.fieldName())) {
+          Object loggedAt = event.get(EventField.EventLoggedAt.fieldName());
+          long loggedAtMs = 0L;
+          if (loggedAt instanceof String) {
+            loggedAtMs = Long.parseLong(
+                event.get(EventField.EventLoggedAt.fieldName()));
+          } else if (loggedAt instanceof Long || loggedAt instanceof Integer) {
+            loggedAtMs = (long) loggedAt;
+          } else if (loggedAt instanceof Timestamp) {
+            loggedAtMs = ((Timestamp) loggedAt).getTime();
+          } else {
+            LOG.warn("Logged at ts unrecognized");
+          }
+          if (loggedAtMs > 0 && loggedAtMs < Integer.MAX_VALUE) {
+            loggedAtMs *= 1000;
+          }
+          if (loggedAtMs != 0) {
+            occurredAtMs = Math.min(occurredAtMs, loggedAtMs);
+          }
+        }
+      } catch (DateTimeParseException | NumberFormatException e) {
+        // skip this event
+        LOG.warn("Skipping event: " + event.get(EventField.EventName.fieldName()));
+      }
+      return occurredAtMs;
+    }
+
+    private static long parseDateTime(String datetime) {
```

Comment:
recommend: I feel we should have a utility class (EventUtils?) with such helper methods to promote code reuse . I see a similar code at  (https://github.com/lyft/beam/blob/bd762d7f35b5b4f458e129b9ba719dcd5ba00299/runners/flink/src/main/java/org/apache/beam/runners/flink/LyftFlinkStreamingPortableTranslations.java#L652)

Reply:
I agree, I was looking to use the parsing logic in EventUtils initially but it looks like its not currently serializable. I believe that would require some work so I've created https://jira.lyft.net/browse/STRMCMP-1109.